Dynamic parameter setting of costmap layers (inflation/static/voxel)

indraneelpatil opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | ros-planning#956 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Turtlebot3 simulation |

---

## Description of contribution in a few bullet points

* Dynamic parameter setting callbacks are added in the layers using suitable mutexes to prevent segfaults
* Decision on whether parameters are settable/non-settable dynamically are taken after referring to these layers in the ROS1 nav stack and warning messages are added wherever parameter setting is rejected
* Appropriate functions are called after dynamic parameter changes (also referring to the ROS1 nav stack)
* Changes are tested on the turtlebot3 simulation
* Suitable test cases are added to test parameter setting

## Description of documentation updates required from your changes

<!--
* Added new parameter, so need to add that to default configs and documentation page
* I added some capabilities, need to document them
-->

---

## Future work that may be required in bullet points

<!--
* I think there might be some optimizations to be made from STL vector
* I see alot of redundancy in this package, we might want to add a function `bool XYZ()` to reduce clutter
* I tested on a differential drive robot, but there might be issues turning near corners on an omnidirectional platform
-->

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

AlexeyMerzlyakov reviewed (commented):
@indraneelpatil, thank you for your work on covering parameters and making it able to be changed on-the-fly!
Some general thoughts. This patch is made for `InflationLayer`, `StaticLayer` and `VoxelLayer`. `ObstacleLayer` and `RangeSensorLayer` are out of scope. So, they could be covered by separate PR. However, since `VoxelLayer` is inherited from `ObstacleLayer` class, they have much in common. So, the main question: should `ObstacleLayer` better be handled by this PR as well (also see some comments from this review), or leave it for next PR?

Also, on this stage, I've made the review for the code. Testcases and coverage will be checked at next step, when the review items will be discussed and met.

indraneelpatil commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
+  }
+
+  if (paramChange_s.inflation_radius || paramChange_s.cost_scaling_factor) {
+    computeCaches();
```
Hey! both computeCaches() and dynamicParametersCallback are using the getMutex() function which is returning a recursive_mutex called access_

So I think the deadlock situation is already taken care of?

AlexeyMerzlyakov replied:
Yes, it is already covered by that needed. That was my oversight.

indraneelpatil commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
+  rcl_interfaces::msg::SetParametersResult result;
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
+
+  struct paramChange
+  {
+    bool inflation_radius : 1;
+    bool cost_scaling_factor : 1;
+    bool enabled : 1;
+    bool inflate_unknown : 1;
+    bool inflate_around_unknown : 1;
+  } paramChange_s{0, 0, 0, 0, 0};
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
+
+  }
+
+  if (paramChange_s.enabled || paramChange_s.inflate_unknown ||
+    paramChange_s.inflate_around_unknown || paramChange_s.inflation_radius ||
+    paramChange_s.cost_scaling_factor)
+  {
+    need_reinflation_ = true;
+  }
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
+      }
+    } else if (param_type == ParameterType::PARAMETER_BOOL) {
+      if (param_name == name_ + "." + "enabled" && enabled_ != parameter.as_bool()) {
+        enabled_ = parameter.as_bool();
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
+      if (param_name == name_ + "." + "inflation_radius" &&
+        inflation_radius_ != parameter.as_double())
+      {
+        inflation_radius_ = parameter.as_double();
+        paramChange_s.inflation_radius = true;
+        cell_inflation_radius_ = cellDistance(inflation_radius_);
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/static_layer.cpp:
```diff
@@ -435,4 +444,46 @@ StaticLayer::updateCosts(
+      if (param_name == name_ + "." + "enabled" && enabled_ != parameter.as_bool()) {
+        enabled_ = parameter.as_bool();
+
+        x_ = y_ = 0;
+        width_ = size_x_;
+        height_ = size_y_;
+        has_updated_data_ = true;
```
Okay done

indraneelpatil commented on nav2_costmap_2d/include/nav2_costmap_2d/static_layer.hpp:
```diff
@@ -178,6 +185,9 @@ class StaticLayer : public CostmapLayer
   nav_msgs::msg::OccupancyGrid::SharedPtr map_buffer_;
+  // Dynamic parameters handler
+  rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr dyn_params_handler_;
+  std::mutex dyn_param_mutex_;
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/voxel_layer.cpp:
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
+
+  }
+
+  matchSize();
```
Done

indraneelpatil commented on nav2_costmap_2d/include/nav2_costmap_2d/voxel_layer.hpp:
```diff
@@ -220,6 +221,18 @@ class VoxelLayer : public ObstacleLayer
+  dynamicParametersCallback(std::vector<rclcpp::Parameter> parameters);
+
+  // Dynamic parameters handler
+  std::mutex mutex_;
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/voxel_layer.cpp:
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
+
+    } else if (param_type == ParameterType::PARAMETER_INTEGER) {
+      if (param_name == name_ + "." + "z_voxels") {
+        size_z_ = parameter.as_int();
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/voxel_layer.cpp:
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
+      } else if (param_name == name_ + "." + "mark_threshold") {
+        mark_threshold_ = parameter.as_int();
+      } else if (param_name == name_ + "." + "combination_method") {
+        combination_method_ = parameter.as_int();
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/voxel_layer.cpp:
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
+      if (param_name == name_ + "." + "enabled") {
+        enabled_ = parameter.as_bool();
+      } else if (param_name == name_ + "." + "footprint_clearing_enabled") {
+        footprint_clearing_enabled_ = parameter.as_bool();
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/voxel_layer.cpp:
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
+      }
+    } else if (param_type == ParameterType::PARAMETER_BOOL) {
+      if (param_name == name_ + "." + "enabled") {
+        enabled_ = parameter.as_bool();
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/voxel_layer.cpp:
```diff
@@ -458,4 +468,60 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
+    if (param_type == ParameterType::PARAMETER_DOUBLE) {
+
+      if (param_name == name_ + "." + "max_obstacle_height") {
+        max_obstacle_height_ = parameter.as_double();
```
List of parameters which are from the parent ObstacleLayer class

- combination_method_
- footprint_clearing_enabled_
- max_obstacle_height_

In the latest commit we have added the mutex locks to the parent ObstacleLayer class so these parameters now seem well protected, so I think we can keep them. Do you insist on their removal?

indraneelpatil commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -108,6 +115,7 @@ InflationLayer::onInitialize()
   need_reinflation_ = false;
   cell_inflation_radius_ = cellDistance(inflation_radius_);
   matchSize();
+
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -407,4 +415,77 @@ InflationLayer::generateIntegerDistances()
+  for (auto parameter : parameters) {
+    const auto & param_type = parameter.get_type();
+    const auto & param_name = parameter.get_name();
+
+
```
Done

indraneelpatil commented:
@AlexeyMerzlyakov Thanks for the great review, I have addressed all comments in the latest commits and ensured test cases are still passing, would love to hear back from your whenever possible

AlexeyMerzlyakov commented on nav2_costmap_2d/test/integration/inflation_tests.cpp:
```diff
@@ -600,3 +601,48 @@ TEST_F(TestNode, testInflation3)
+ */
+TEST_F(TestNode, testDynParamsSet)
+{
+  auto node = std::make_shared<rclcpp_lifecycle::LifecycleNode>("inflation_dyn_param_test");
```
This parameter is unused. Why do we need to keep it?

indraneelpatil replied:
Aaah, thanks great spot!

AlexeyMerzlyakov commented on nav2_costmap_2d/test/integration/obstacle_tests.cpp:
```diff
@@ -405,5 +406,103 @@ TEST_F(TestNode, testRaytracing) {
+ */
+TEST_F(TestNode, testDynParamsSetVoxel)
+{
+  auto node = std::make_shared<rclcpp_lifecycle::LifecycleNode>("voxel_dyn_param_test");
```
Again, this looks like an unnecessary code.

indraneelpatil replied:
Done!

AlexeyMerzlyakov commented on nav2_costmap_2d/test/integration/obstacle_tests.cpp:
```diff
@@ -405,5 +406,103 @@ TEST_F(TestNode, testRaytracing) {
+    costmap->get_node_graph_interface(),
+    costmap->get_node_services_interface());
+
+  auto results = parameter_client->set_parameters_atomically(
+  {
+    rclcpp::Parameter("voxel_layer.combination_method", 0),
+    rclcpp::Parameter("voxel_layer.mark_threshold", 1),
+    rclcpp::Parameter("voxel_layer.unknown_threshold", 10),
+    rclcpp::Parameter("voxel_layer.z_resolution", 0.4),
+    rclcpp::Parameter("voxel_layer.origin_z", 1.0),
+    rclcpp::Parameter("voxel_layer.z_voxels", 14),
+    rclcpp::Parameter("voxel_layer.max_obstacle_height", 4.0),
+    rclcpp::Parameter("voxel_layer.footprint_clearing_enabled", false),
+    rclcpp::Parameter("voxel_layer.enabled", false)
+  });
```
One corner case is not being covered:
![Screenshot 2022-01-28 at 16-57-20 LCOV - dynamic_param_1 info - plugins voxel_layer cpp](https://user-images.githubusercontent.com/60094858/151562359-cc139b3d-3b74-4f65-8797-7e18f06c8b9e.png)

indraneelpatil replied:
Done

AlexeyMerzlyakov commented on nav2_costmap_2d/test/integration/obstacle_tests.cpp:
```diff
@@ -405,5 +406,103 @@ TEST_F(TestNode, testRaytracing) {
+ */
+TEST_F(TestNode, testDynParamsSetStatic)
+{
+  auto node = std::make_shared<rclcpp_lifecycle::LifecycleNode>("static_dyn_param_test");
```
The same comment (extra-code).

indraneelpatil replied:
Done!

AlexeyMerzlyakov commented on nav2_costmap_2d/test/integration/obstacle_tests.cpp:
```diff
@@ -405,5 +406,103 @@ TEST_F(TestNode, testRaytracing) {
+    costmap->get_node_graph_interface(),
+    costmap->get_node_services_interface());
+
+  auto results = parameter_client->set_parameters_atomically(
+  {
+    rclcpp::Parameter("static_layer.transform_tolerance", 1.0),
+    rclcpp::Parameter("static_layer.enabled", false)
+  });
```
Some corner cases are not being covered (as follows):
![Screenshot 2022-01-28 at 16-56-32 LCOV - dynamic_param_1 info - plugins static_layer cpp](https://user-images.githubusercontent.com/60094858/151562669-1921fe91-9a4d-4603-b117-97d20388b281.png)

indraneelpatil replied:
Done!

AlexeyMerzlyakov reviewed (commented):
Sorry for my late reply due to some sick last week.
Regarding the change, this seems is much better. Some few strokes before we go.

Regarding separation of `ObstacleLayer`, I personally would like to concatenate these two into one PR, but also don't mind for your option. The change for `ObstacleLayer` could take a new Observation API, so this should be paid an attention in the next step as separate PR. I am personally OK abut it too.

indraneelpatil commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -407,4 +417,65 @@ InflationLayer::generateIntegerDistances()
+InflationLayer::dynamicParametersCallback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+  static bool need_cache_recompute = false;
+
+  rcl_interfaces::msg::SetParametersResult result;
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
```
Sounds good!

indraneelpatil commented on nav2_costmap_2d/plugins/voxel_layer.cpp:
```diff
@@ -458,4 +468,68 @@ void VoxelLayer::updateOrigin(double new_origin_x, double new_origin_y)
+VoxelLayer::dynamicParametersCallback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+  static bool resize_map_needed = false;
+  rcl_interfaces::msg::SetParametersResult result;
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -407,4 +417,65 @@ InflationLayer::generateIntegerDistances()
+{
+  static bool need_cache_recompute = false;
+
+  rcl_interfaces::msg::SetParametersResult result;
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/static_layer.cpp:
```diff
@@ -435,4 +444,47 @@ StaticLayer::updateCosts(
+StaticLayer::dynamicParametersCallback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+  rcl_interfaces::msg::SetParametersResult result;
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
```
Done

indraneelpatil commented on nav2_costmap_2d/include/nav2_costmap_2d/voxel_layer.hpp:
```diff
@@ -51,6 +51,7 @@
 #include <message_filters/subscriber.h>
 #include <nav2_costmap_2d/obstacle_layer.hpp>
 #include <nav2_voxel_grid/voxel_grid.hpp>
+#include <mutex>
```
Done

indraneelpatil commented:
@AlexeyMerzlyakov I covered the latest review! Hoping to hear back from you soon!

AlexeyMerzlyakov commented on nav2_costmap_2d/test/integration/obstacle_tests.cpp:
```diff
@@ -405,5 +406,107 @@ TEST_F(TestNode, testRaytracing) {
+  costmap->on_deactivate(rclcpp_lifecycle::State());
+  costmap->on_cleanup(rclcpp_lifecycle::State());
+  costmap->on_shutdown(rclcpp_lifecycle::State());
+
```
Extra blank

indraneelpatil replied:
Done

AlexeyMerzlyakov commented on nav2_costmap_2d/test/integration/inflation_tests.cpp:
```diff
@@ -600,3 +601,46 @@ TEST_F(TestNode, testInflation3)
+  costmap->on_deactivate(rclcpp_lifecycle::State());
+  costmap->on_cleanup(rclcpp_lifecycle::State());
+  costmap->on_shutdown(rclcpp_lifecycle::State());
+
```
Extra blank

indraneelpatil replied:
Done

AlexeyMerzlyakov reviewed (approved):
All previous items were met. Please remove extra blank lines (pointed out where) and ready for merge.

indraneelpatil commented on nav2_costmap_2d/plugins/static_layer.cpp:
```diff
@@ -435,4 +444,48 @@ StaticLayer::updateCosts(
+StaticLayer::dynamicParametersCallback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+
```
Done

indraneelpatil commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -407,4 +417,65 @@ InflationLayer::generateIntegerDistances()
+InflationLayer::dynamicParametersCallback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+
```
Done

indraneelpatil commented on nav2_costmap_2d/test/integration/obstacle_tests.cpp:
```diff
@@ -405,5 +406,107 @@ TEST_F(TestNode, testRaytracing) {
+
+}
+
+
```
Done

indraneelpatil commented on nav2_costmap_2d/test/integration/obstacle_tests.cpp:
```diff
@@ -405,5 +406,107 @@ TEST_F(TestNode, testRaytracing) {
+  costmap->on_deactivate(rclcpp_lifecycle::State());
+  costmap->on_cleanup(rclcpp_lifecycle::State());
+  costmap->on_shutdown(rclcpp_lifecycle::State());
 
```
Done

indraneelpatil commented on nav2_costmap_2d/test/integration/inflation_tests.cpp:
```diff
@@ -600,3 +601,46 @@ TEST_F(TestNode, testInflation3)
   ASSERT_EQ(countValues(*costmap, nav2_costmap_2d::INSCRIBED_INFLATED_OBSTACLE), 4u);
 }
+
+
```
Done

indraneelpatil commented:
@AlexeyMerzlyakov Removed blank lines, thanks!

AlexeyMerzlyakov commented:
@SteveMacenski, I think, we could merge this.

SteveMacenski commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -113,6 +120,7 @@ InflationLayer::onInitialize()
 void
 InflationLayer::matchSize()
 {
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
```
My understanding of the mutex in the inflation layer is that it is for locking in the inflation resource (see the doxygen comment on it) and the previous implementation was only locking the update / values in the inflation cost data structure. It feels like locking it also for the parameters is abusing the role here a little.

 Would it make more sense to have these variables as atomics instead?

SteveMacenski replied:
Good point. Maybe its better to use another mutex then -- or do you think its OK that we're kind of abusing this mutex? Its not really for the same thing anymore. At bare minimum, the description of the mutex needs updating.

AlexeyMerzlyakov replied:
For InflationLayer, standard mutex was chosen to cover `inflation_radius_` and `need_reinflation_` variables shared usage by `matchSize()`, `updateBounds()`, `onFootprintChanged()` and `dynamicParametersCallback()`. However, introducing of new lock for them will cause to have 2 mutexes in `updateCosts()` + `computeCaches()` or `dynamicParametersCallback()` which is also accessing `cell_inflation_radius_` parameter used by already covered by standard mutex routines. Therefore, I've suggested to use one mutex for everything for simplicity.

SteveMacenski replied:
OK, I think that's fine

SteveMacenski commented on nav2_costmap_2d/plugins/obstacle_layer.cpp:
```diff
@@ -455,6 +455,7 @@ ObstacleLayer::updateFootprint(
   double * max_x,
   double * max_y)
 {
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
```
Why is this costmap layer not having any dynamic parameters? I see the voxel does, which derives from this, and would implicitly add those dynamic parameters to both sets of layers at once

indraneelpatil replied:
Yes, so we originally started off the PR with Obstacle Layer out of scope of these changes (ref Alexey's comments above) but then we realised that it wasnt possible to make voxel dynamic without also making obstacle layer dynamic.

But then we decided that since this PR is already pretty big, reviewing it and keeping it in sync with master is becoming a problem so we think its best if we target ObstacleLayer in the next PR. Is this okay?

SteveMacenski replied:
OK!

indraneelpatil commented:
@SteveMacenski  Yes we used the turtlebot3 simulation to try all these parameter changes out!

SteveMacenski commented on nav2_costmap_2d/plugins/inflation_layer.cpp:
```diff
@@ -407,4 +417,64 @@ InflationLayer::generateIntegerDistances()
+
+  }
+
+  if (need_cache_recompute) {
```
Wouldn't it be better at this point to call `matchSize` which will handle `computeCaches()` and do all the `cellDistance()` stuff for you? That way the if/else conditions above can be more simple.

indraneelpatil replied:
I guess you are right, the decision of computeCache() or matchSize() I just made based on the ROS1 Nav stack https://github.com/ros-planning/navigation/blob/noetic-devel/costmap_2d/plugins/inflation_layer.cpp

Practically I think the only difference between the two approaches is that through matchSize() we are additionally resetting the resolution_ and seen vectors. But it doesnt seem like that causes any issues. Pushing the change.

SteveMacenski reviewed (approved):
@AlexeyMerzlyakov merge if you're happy with it, I am happy with it. You took the lead on it so I'll let you hit the button :-) 

@indraneelpatil thanks for the help and the repeated review cycles! I appreciate this feature so much and we're crazy close to being done with all this dynamic parameter infrastructure! I definitely look foward to the obstacle layer's.

AlexeyMerzlyakov commented:
Also, looks good to me. @indraneelpatil, thank you for your well-made work! Looking forward for next part of dynamic parameters PR.
